File: Java/p401.java
import java.util.Scanner;

public class p401 {
    public static void main (String [] args) {
        Scanner input = new Scanner(System.in);
        while (input.hasNext()) {
            String kata = input.next();
            boolean palindromeCheck = isPalindrome(kata, 0, kata.length()-1);
            boolean mirrorCheck = isMirrored(kata, 0, kata.length()-1);
            if (palindromeCheck) {
                if (mirrorCheck) {
                    System.out.println(kata + " -- is a mirrored palindrome.\n");
                } else {
                    System.out.println(kata + " -- is a regular palindrome.\n");
                }
            } else {
                if (mirrorCheck) {
                    System.out.println(kata + " -- is a mirrored string.\n");
                } else {
                    System.out.println(kata + " -- is not a palindrome.\n");
                }
            }
        }
    }
    
    static boolean isPalindrome (String kata, int low, int high) {
        if (low > high) return true;
        else {
            if (kata.charAt(low) == kata.charAt(high)) return isPalindrome(kata, low += 1, high -= 1);
            else return false;
        }
    }
    
    static boolean isMirrored (String kata, int low, int high) {
        if (low > high) return true;
        else {
            if (cekMirror(kata.charAt(low),kata.charAt(high)))
                return isMirrored(kata, low += 1, high -= 1);
            else return false;
        }
    }
   
    static boolean cekMirror (char asal, char pembanding) {
        if ((asal == 'E' && pembanding == '3') || (asal == '3' && pembanding == 'E')) return true;
        else if ((asal == 'J' && pembanding == 'L') || (asal == 'L' && pembanding == 'J')) return true;
        else if ((asal == 'S' && pembanding == '2') || (asal == '2' && pembanding == 'S')) return true;
        else if ((asal == 'Z' && pembanding == '5') || (asal == '5' && pembanding == 'Z')) return true;
        else if (asal == 'A' || asal == 'H' || asal == 'I' || asal == 'M' || asal == 'O'
            || asal == 'T' || asal == 'U' || asal == 'V' || asal == 'W' || asal == 'X'
            || asal == 'Y' || asal == '1' || asal == '8') {
                if (asal == pembanding) return true;
                else return false;
            }
        else return false;
    }
}

File: C,C++/p401.cpp
#include <iostream>
#include <string>

using namespace std;

bool isPalindrome(string, int, int);
bool isMirrored(string);
string reverse(string);

int main () {
    string kata;
    while (getline(cin, kata)) {
        int length = kata.length();
        if (isPalindrome(kata, 0, length-1)) {
            if (isMirrored(kata)) {
                cout << kata <<" -- is a mirrored palindrome." << endl;
            } else {
                cout << kata <<" -- is a regular palindrome." << endl;
            }
        } else {
            if (isMirrored(kata)) {
                cout << kata <<" -- is a mirrored string." << endl;
            } else {
                cout << kata <<" -- is not a palindrome." << endl;
            }
        }
    }
   return 0;
}

bool isPalindrome(string kata, int low, int high) {
    if (low > high) {
        return true;
    } else {
        if (kata.at(low) == kata.at(high)) return isPalindrome(kata, low += 1, high -= 1);
        else return false;
    }
}

bool isMirrored(string kata) {
    string reverseWord = reverse(kata);
    string temp = "";
    int length = reverseWord.size()-1;
    for (int i = length; i >= 0; i--) {
    	temp.push_back(reverseWord.at(i));
	}
//    cout << temp << " " << endl;
//    cout << kata.compare(reverseWord) << endl;
    if (temp.compare(kata) == 0) return true;
    else return false;
}

string reverse(string kata) {
    string hasil = "";
    for (int i = 0; kata[i] != '\0'; i++) {
        if (kata.at(i) == 'E') {
            hasil.push_back('3');
        } else if (kata.at(i) == '3') {
            hasil.push_back('E');
        } else if (kata.at(i) == 'J') {
            hasil.push_back('L');
        } else if (kata.at(i) == 'L') {
            hasil.push_back('J');
        } else if (kata.at(i) == 'S') {
            hasil.push_back('2');
        } else if (kata.at(i) == '2') {
            hasil.push_back('S');
        } else if (kata.at(i) == 'Z') {
            hasil.push_back('5');
        } else if (kata.at(i) == '5') {
            hasil.push_back('Z');
        } else {
            hasil.push_back(kata.at(i));
        }
    }
    return hasil;
}
